### PR TITLE
Fix orientation of subframe offset before applying to motion plans

### LINKED ISSUE #2879: PILZ Planning with attached collision object subframe
### Description

Trying to use an attached collision object subframe as an end-effector link for planning. It works as expected with OMPL, but with Pilz it seems like the transformation of the subframe is just applied w.r.t the world frame and not w.r.t the attached object pose.

### Your environment
* ROS Distro: [Noetic]
* OS Version: e.g. Ubuntu 20.04
* Master

### Steps to reproduce
- add `<arg name="planning_adapters" value="default_planner_request_adapters/ResolveConstraintFrames" />` to both pilz and ompl `planning_pipeline.launch.xml`
- set `move_group.setEndEffectorLink("cylinder/tip")` as in the subframe tutorial, e.g. shifted from TCP frame along `z=0.1`.
- Move to a given target with OMPL and compare with PILZ

### Expected behaviour
- The robot moves to the same target with same IK frame with PILZ and OMPL

### Actual behaviour
- The robot moves to the target with a different IK frame, e.g. OMPL moves to expected pose, but PILZ moves ``0.2`` below desired pose.

### PR DESCRIPTION
**Pull request by rivelin-tom**
### Description

This PR fixes #2879 by reorienting the subframe offset applied to a goal pose in the PTP planner, so that it is applied in the correct direction. It also adds the same offset computation to LIN and CIRC motions, allowing them to work with subframes.

The video below demonstrates the fix by showing that the goal point planned to by pilz PTP is now the same as OMPL.

https://user-images.githubusercontent.com/86778266/135306658-ac17ea25-7461-47ad-8370-7b44f8705931.mp4

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/) (Not applicable)
- [x] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes (Not applicable)
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [x] Include a screenshot if changing a GUI (Not applicable)
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by simonschmeisser** [comment]
Could those calculations be moved to a common location? Are they maybe even useful outside of `pilz`?

**Comment by robertjbush** [comment]
@simonschmeisser While they are probably useful, those calculations are a specific fix for this `pilz` bug and this bug only. I gave @aa-tom the mandate to fix this bug and not add enhancements to the MoveIt! codebase. We'd appreciate some help from the MoveIt! community if your suggestion is desired, otherwise it would be good to merge this before the 1.1.6 MoveIt! release.

**Comment by robertjbush** [comment]
Also fixes #2884

**Comment by aa-tom** [comment]
Agree that it'd be nice to make this code common if possible. The most reasonable place seems like the ResolveConstraintFrames plugin, which calculates the subframe offsets before the plan request reaches the planner. I have a couple of concerns though:

- I'm not too familiar with the plugin, so would want to make sure that there isn't a specific reason why the offsets aren't already being oriented using the goal pose there. @rhaschke could you weigh in here? I think you originally developed the ResolveConstraintFrames plugin?

- I think making the change in the plugin could have unintended effects in other planners. OMPL was never broken, which suggests to me that there's already some kind of offset computation going on in there. I'm worried that, without a thorough investigation of which planners would need to change in response to putting the offset computation in the plugin, we could end up breaking more plans than we fix. Maybe worth its own issue though?

**Review by jschleicher — COMMENTED** [review]
Thank you very much for your investigation!

Other planners rely on the `KinematicConstraint` classes, which are probably designed for sampling. This is a fundamental differnce of OMPL - here we just use a single constraint. I would prefer a more general approach, but currently am not aware of a simple solution.

To avoid code-duplication, you could still move the offset part into trajectory_functions.cpp or similar.

Note that for a circle, you probably also have to add the offset to the circ_path_point (interim or center): https://github.com/ros-planning/moveit/blob/049be73b8343d67ffde400fed9ed4c387c6ea9e5/moveit_planners/pilz_industrial_motion_planner/src/trajectory_generator_circ.cpp#L183

File: moveit_planners/pilz_industrial_motion_planner/src/trajectory_generator_ptp.cpp [hunk 1]
```diff
@@ -239,10 +239,14 @@ void TrajectoryGeneratorPTP::extractMotionPlanInfo(const planning_scene::Plannin
...
-    p.z -= req.goal_constraints.at(0).position_constraints.at(0).target_point_offset.z;
-
+    Eigen::Vector3d offset_in_world;
+    tf2::fromMsg(req.goal_constraints.at(0).position_constraints.at(0).target_point_offset, offset_in_world);
+    Eigen::Quaterniond goal_orientation;
+    tf2::fromMsg(req.goal_constraints.at(0).orientation_constraints.at(0).orientation, goal_orientation);
+    Eigen::Vector3d offset_in_goal = goal_orientation * offset_in_world;
+    p.x -= offset_in_goal[0];
+    p.y -= offset_in_goal[1];
+    p.z -= offset_in_goal[2];
```
[the remark this answers is not in the document]
Comment (aa-tom) [reply moveit_planners/pilz_industrial_motion_planner/src/trajectory_generator_ptp.cpp:249]: Thanks for the advice! Will get those tasks done tomorrow:
- [ ] Move common offset transformation into `trajectory_functions.cpp`
- [ ] Apply offset to center for CIRC motions
- [ ] Tidy up Eigen computations

**Comment by aa-tom** [comment]
@jschleicher @rhaschke @v4hn Any ideas why the build might be failing?

**Comment by aa-tom** [comment]
Looks like the Windows build failure is the same as this https://github.com/ros-planning/moveit/pull/2892#issuecomment-933032054

Unsure why so many tests would be failing provided they aren't all using offsets and aren't checking for specific end poses

**Comment by v4hn** [comment]
> Unsure why so many tests would be failing provided they aren't all using offsets and aren't checking for specific end poses

I just triggered re-runs to rule out a system issue, but you probably broke something and will have to look into it. :)
Just run the specific failing tests locally to debug.

File: moveit_planners/pilz_industrial_motion_planner/include/pilz_industrial_motion_planner/trajectory_functions.h [hunk 2]
```diff
@@ -220,3 +220,10 @@ bool isStateColliding(const bool test_for_self_collision, const planning_scene::
 }  // namespace pilz_industrial_motion_planner
 
 void normalizeQuaternion(geometry_msgs::Quaternion& quat);
+
+/**
+ * @brief Fetches the pose defining a constraint with any offset applied.
+ * @param constraint
+ * @return
+ */
+Eigen::Isometry3d getConstraintPose(moveit_msgs::Constraints constraint);
```
Comment (v4hn) [inline moveit_planners/pilz_industrial_motion_planner/include/pilz_industrial_motion_planner/trajectory_functions.h:229]:
```suggestion
Eigen::Isometry3d getConstraintPose(const moveit_msgs::Constraints& constraint);
```

**Comment by jschleicher** [comment]
@aa-tom You can run the tests locally with more output using for example `rostest --text pilz_industrial_motion_planner integrationtest_command_planning_frankaemika_panda.test`
Which gives an output of
```
[ INFO] [1633345214.152003125]: Received new planning service request...
[ INFO] [1633345214.152090320]: The timeout for planning must be positive (0.000000 specified). Assuming one second instead.
[ INFO] [1633345214.152116846]: Generating CIRC trajectory...
[ERROR] [1633345214.152365149]: Exception caught: 'vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)'
```
so maybe some check for empty constraint lists got lost?

**Comment by aa-tom** [comment]
Thanks. Will run the tests locally and take a look this afternoon

File: moveit_planners/pilz_industrial_motion_planner/include/pilz_industrial_motion_planner/trajectory_functions.h (hunk 2 above)
Comment (aa-tom) [reply moveit_planners/pilz_industrial_motion_planner/include/pilz_industrial_motion_planner/trajectory_functions.h]: Done :+1:

**Comment by rhaschke** [comment]
@rivelin-tom, what is your progress on https://github.com/ros-planning/moveit/pull/2890#issuecomment-949838481? This PR is currently blocking the next Noetic release as your colleague? asked for merging this before: https://github.com/ros-planning/moveit/pull/2890#issuecomment-931113662

**Comment by rhaschke** [comment]
Ping @rivelin-tom.

File: moveit_planners/pilz_industrial_motion_planner/src/trajectory_functions.cpp [hunk 3]
```diff
@@ -585,3 +585,18 @@ void normalizeQuaternion(geometry_msgs::Quaternion& quat)
   tf2::fromMsg(quat, q);
   quat = tf2::toMsg(q.normalize());
 }
+
+Eigen::Isometry3d getConstraintPose(const moveit_msgs::Constraints& constraint, const geometry_msgs::Vector3& offset,
+                                    const geometry_msgs::Quaternion& orientation)
+{
+  geometry_msgs::Pose pose;
+  pose.position = constraint.position_constraints.front().constraint_region.primitive_poses.front().position;
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_planners/pilz_industrial_motion_planner/src/trajectory_functions.cpp:593]: I addressed this feedback in https://github.com/rivelinrobotics/aa-pilz/pull/5. Please merge this and provide unit tests. Thanks!

**Comment by simonschmeisser** [comment]
@rhaschke should this be closed since you merged it manually?

**Comment by rhaschke** [comment]
> Should this be closed since you merged it manually?

No, I'm still waiting for unit tests to be added... Unfortunately, there was no activity from @rivelin-tom since 15 days.
I took the time to fix the code as suggested, merge the bug fix, and thus unblock the release.

**Comment by robertjbush** [comment]
@rhaschke Apologies for the delay but @rivelin-tom has had to implement some features and enhancements for us. We've not forgotten about it, just had some urgent and critical work to get out the way first.

**Comment by rivelin-tom** [comment]
@rhaschke Thanks a bunch for getting it ready for the release and apologies for the delay. I'm super busy at the moment, so it'll be a little while before I'm able to pick this back up and write the tests. I do intend to write them!

**Comment by gautz** [comment]
Is this replaced by [e30c6fa](https://github.com/ros-planning/moveit/commit/e30c6fa853dbc699c0cc84e420626000eefe3212)?
Should this PR be further extended to correctly transform the CIRC interim frame when setIKframe() is set to an attached object subframe in MTC ? See the issue [described here
](https://github.com/ros-planning/moveit/issues/2250#issuecomment-970324091).

see also https://github.com/ros-planning/moveit_task_constructor/issues/191 and https://github.com/ros-planning/moveit_task_constructor/pull/304

**Comment by v4hn** [comment]
@rivelin-tom Is it still realistic to wait for your tests here? I know you want to do them, but you are obviously busy with other things.

> Should this PR be further extended

No, feel free to open separate PRs if you have patches.

**Comment by rivelin-tom** [comment]
@v4hn I think it'd be best to close this off for now to avoid delaying longer. I believe[ it was already merged with master as part of the 1.1.6 release](https://github.com/ros-planning/moveit/blob/master/moveit_planners/pilz_industrial_motion_planner/CHANGELOG.rst)? Happy for this PR to simply be closed if that's the case

**Comment by rhaschke** [comment]
Yes, the fix was already merged. However, this PR is still open as a reminder that you, @rivelin-tom, promised to provide some unit tests. Of course, you can close here and open a new PR for those tests.

**Comment by rivelin-tom** [comment]
Sorry, was away for Christmas. I will do that, then. Thanks for the help implementing the changes :+1:

**Pull request closed without merging**